src/mrinufft/operators/interfaces/torchkbnufft.py

```diff
@@ -0,0 +1,202 @@
+            samples.astype(np.float32, copy=False), normalize="pi"
+        )
+        self.samples = samples.transpose(1, 0)
+        self.samples = torch.tensor(samples)
+
+        if density is True:
+            self.density = torchnufft.calc_density_compensation_function(
+                ktraj=self.samples, im_size=shape, num_iterations=15
+            )
+        elif density is False:
+            self.density = None
+        elif (
+            torch.is_tensor(density)
+            or isinstance(density, np.ndarray)
+            or isinstance(density, cp.ndarray)
+        ):
+            self.density = density
+        else:
+            raise ValueError(
+                "argument `density` of type" f"{type(density)} is invalid."
+            )
+
+        if smaps is None:
+            self.smaps = None
+        elif isinstance(smaps, torch.Tensor):
+            self.smaps = smaps
+        elif isinstance(smaps, np.ndarray) or isinstance(smaps, cp.ndarray):
+            self.smaps = torch.tensor(smaps)
+        else:
+            raise ValueError("argument `smaps` of type" f"{type(smaps)} is invalid")
+
+    @with_torch
+    def op(self, data, ksp=None):
+        """Forward operation.
+
+        Parameters
+        ----------
+        data: Tensor
+
+        Returns
+        -------
+        Tensor: Non-uniform Fourier transform of the input image.
+        """
+        ktraj = self.samples
+        smaps = self.smaps
+
+        B, C, XYZ = self.n_batchs, self.n_coils, self.shape
+        if not B:
+            B = 1
+        if not C:
+            C = 1
+        data = data.reshape((B, 1 if self.uses_sense else C, *XYZ))
+
+        if ktraj.shape[0] != data.shape[0]:
+            ktraj = ktraj.permute(1, 0)
+        if smaps is not None:
+            smaps = smaps.to(data.dtype)
+        kdata = self._tkb_op.forward(image=data, omega=ktraj, smaps=smaps)
+        kdata /= self.norm_factor
+
+        return self._safe_squeeze(kdata)
+
+    @with_torch
+    def adj_op(self, data, coeffs=None):
+        """Backward Operation.
+
+        Parameters
+        ----------
+        data: Tensor
+
+        Returns
+        -------
+        Tensor
+        """
+        ktraj = self.samples
+        smaps = self.smaps
+        B, C, K, XYZ = self.n_batchs, self.n_coils, self.n_samples, self.shape
+        if not B:
+            B = 1
+        if not C:
+            C = 1
+        data = data.reshape((B, C, K))
+
+        if ktraj.shape[0] != data.shape[0]:
+            ktraj = ktraj.permute(1, 0)
+        if self.density:
+            data = data * self.density
+
+        if smaps is not None:
+            smaps = smaps.to(data.dtype)
+
+        img = self._tkb_adj_op.forward(data=data, omega=ktraj, smaps=smaps)
+        img = img.reshape((B, 1 if self.uses_sense else C, *XYZ))
+        img /= self.norm_factor
+
+        return self._safe_squeeze(img)
+
+    @with_torch
+    def data_consistency(self, data, obs_data):
```

Comment:
with_torch does not support the two arguments, best is to resort to manual checking here.

Reply:
I don't know if i can do it like this , I left @with_torch to control "data" type and I control "obs_data" in the present function .